scrapypuppeteer/response.py

```diff
@@ -70,14 +77,65 @@ def follow(
             **kwargs,
         )
 
+    def follow_all(
+        self,
+        actions=None,
+        close_page: bool = True,
+        accumulate_meta: bool = False,
+        css=None,
+        xpath=None,
+        **kwargs,
+    ) -> Generator[PuppeteerRequest, None, None]:
+        """
+        Execute actions in the same context but in other browser pages.
+        Only one of `actions`, `css`, or `xpath` must be specified.`
+        Note that original page from which the method was called lasts unaffected.
+
+        :param actions: iterable of PuppeteerActions or selectors
+        :param close_page: whether to close page after request completion
+        :param accumulate_meta: whether to accumulate meta from response
+        :param css: selector
+        :param xpath: selector
+        :return: Iterable[PuppeteerRequest]
+        """
 
-class PuppeteerHtmlResponse(PuppeteerResponse):
+        # Probably, we should ban any PuppeteerAction in `actions` except GoTo
+        arguments = [x for x in (actions, css, xpath) if x is not None]
+        if len(arguments) != 1:
+            raise ValueError(
+                "Please supply exactly one of the following arguments: actions, css, xpath"
+            )
+        if not actions:
+            if css:
+                actions = self.css(css)
+            if xpath:
+                actions = self.xpath(xpath)
+
+        page_id = self.page_id  # Substitution of page_id in order to create new page
+        self.page_id = None
+
+        yield from (
+            self.follow(
+                action,
+                close_page=close_page,
+                accumulate_meta=accumulate_meta,
+                **kwargs,
+            )
+            for action in actions
+        )
+
+        self.page_id = page_id
```

Comment:
I dislike this pattern because it modifies the response during iteration and allows constructions like
```
requests_iter = response.follow_all(links)
assert response.page_id is not None
next_request = next(requests_iter, None)
assert response.page_id is None
```

I suggest either passing page_id=None to self.follow, or collecting following requests to list first.

Reply:
First option is not possible right now because in `self.follow` we determine `page_id` as `page_id = None if self.puppeteer_request.close_page else self.page_id`.
I modified my solution so changing `self.page_id` inside `follow_all` method is not visible to user.
Still, it modifies the response during iteration. I like the way of providing `page_id=None` to self.follow but currently this raises an error. Probably, it's a bug.
I suggest to solve the problem within the PR related to [this comment](https://github.com/ispras/scrapy-puppeteer-service/pull/53#discussion_r1771159218). We will properly validate arguments not only in the service but also in the client part.